### PR TITLE
Converted to TypeScript

### LINKED ISSUE #232: Update to typescript
Since we already have typings and use babel anyway, it shouldn't be difficult to just switch to TypeScript entirely.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #232.

Converts `.js` files to `.ts`/`.tsx`. 

* Removes `index.d.ts` and replaces its type definitions with a new `types.ts`, which is `*` exported from `index.ts`
* Removes Babel and `types-tester` because TypeScript does both transpilation and type checking
    * Adds `.js.map` and `.d.ts` files to `lib/`
* Removes `jsnext:main` because of https://github.com/jsforum/jsforum/issues/5 and to keep things simpler
* Moves `defaultProps` and `propTypes` into `static` members of their classes for type safety

I could add Babel back in with the TypeScript preset and/or add another compile target for `jsnext:main`/`pkg.module`, but opted not to for now to keep things simple. Let me know if you'd like something!

You should also definitely look at the different output in `lib/`. TypeScript has a bit simpler of an output target, and although this still works for me, it might not work in your browser/babel combination.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,17 +4,16 @@
...
+    "build": "tsc",
     "clean": "rimraf lib",
-    "format": "eslint --fix '{src,test,example}/**/*.{ts,js}'",
-    "lint": "eslint '{src,test,example}/**/*.{ts,js}'",
+    "format": "eslint --fix '{src,example}/**/*.{ts,tsx}'",
+    "lint": "eslint '{src,example}/**/*.{ts,tsx}'",
     "prepublishOnly": "npm run lint && npm run test && npm run build",
     "test": "npm run test:typescript",
-    "test:typescript": "typings-tester --dir test/typescript"
+    "test:typescript": "tsc -p test/typescript"
```
Comment (domoritz) [inline package.json:16]: we can probably remove this now

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export function processSize(size?: number | string) {
```
Comment (domoritz) [inline src/utils.ts:1]: Shouldn't this be

```suggestion
export function processSize(size: number | string) {
```

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "moduleResolution": "node",
+    "noFallthroughCasesInSwitch": true,
+    "noImplicitThis": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "skipLibCheck": true,
+    "sourceMap": true,
+    "target": "es2015"
```
Comment (domoritz) [inline tsconfig.json:18]: Could using es6 cause problems with older browsers? Maybe we still want to use babel to generate es5 and es6 compliant code? Or we just run babel after transpiling with tsc to get es5 compliant code (I do this in https://github.com/vega/vega-embed/blob/f9450f402da127fa3369ad12606b0964298c3a65/package.json#L69).

Comment (JoshuaKGoldberg) [reply tsconfig.json:19]: I went with the original system to keep to just one output directory (`lib/`) and no intermediaries.

I also noticed that `Object.assign` doesn't get polyfilled in the output `lib/editor.js` on `master`, which is breaking this package on IE 11 😱 ... #293

Comment (JoshuaKGoldberg) [reply tsconfig.json:19]: Actually, @domoritz, once #294's fix to #293 lands, this PR can have the target be `es5`, which would let the output run on browsers as old as IE 11. That'd be a bit cleaner IMO than continuing to use Babel. Thoughts?

Comment (domoritz) [reply tsconfig.json:19]: Sounds good. I merged #294

**Comment by domoritz** [comment]
Thank you. This looks great. Can you resolve the merge conflict and then I can merge.

**Pull request merged**